docs: ESM update (#13191)
facebook/jest: docs/ECMAScriptModules.md

// jest === import.meta.jest => true
```

Additionally, since ESM evaluates static `import` statements before looking at the code, hoisting on `jest.mock` calls that happens in CJS modules won't work in ESM. To `mock` modules in ESM, you need to use dynamic `import()` after `jest.mock` calls to load the mocked modules, same applies to modules which have to load the mocked modules.

> ```js
> // - main.cjs

Reviewer: Here and other places:

```suggestion
```js  title="main.cjs"
```

const { BrowserWindow, app } = require('electron');

// etc.

module.exports = { example };
```

> ```js
> // - main.test.js

Reviewer: Here and other places:

```suggestion
```js  title="main.test.js"
```

import { jest } from '@jest/globals';

jest.mock('electron', () => ({
  app: {
    on: jest.fn(),
    whenReady: jest.fn(() => Promise.resolve()),
  },
  BrowserWindow: jest.fn().mockImplementation(() => ({
    // partial mocks.
  }))
}));

> const { BrowserWindow } = (await import('electron')).default;

Reviewer: Hm.. Felt like `require` look better for this case. That’s just style, I’m fine with anything what is working (;

Author: I'm fine to go with either style there, just that if we go `require` for `electron` then we should do the same with `main.cjs` as Electron still doesn't work with ESM yet. Tbh I'm feeling a bit strange to use something that only works with CJS as an example here but just that I was working on a Electron project in TypeScript so I used it in my minimal repo for my issue 🤔

// jest === import.meta.jest => true
```

> Additionally, since ESM evaluates static `import` statements before looking at the code, hoisting on `jest.mock` calls that happens in CJS modules won't work in ESM. To `mock` modules in ESM, you need to use dynamic `import()` after `jest.mock` calls to load the mocked modules, same applies to modules which have to load the mocked modules.

Reviewer 2: why do we talk about "dynamic `import()`, and then go on to use `require`?

Author: I've made the change to also mention `require` there, with the example now showing both `require` and dynamic `import()` usage. Personally I prefer `require` for CJS modules but I think it's better to show both on the docs so the users aren't going to think that either is more preferable.

```js title="main.cjs"
const { BrowserWindow, app } = require('electron');

// etc.

module.exports = { example };
```

> ```js title="main.test.cjs"

Reviewer 2: ```suggestion
```js title="main.test.js"
```

import { createRequire } from 'node:module';
import { jest } from '@jest/globals';

const require = createRequire(import.meta.url);

> jest.mock('electron', () => ({

Reviewer 2: note that this only works because you use `require` in "production" code - if that was using `import` (which I would think everybody reading these docs are doing) then it won't work. So I think we should have the unstable API first, then go on to say that an alternative is to use CJS wrappers and mock them

Author: You do have a point there. Lemme rearrange the order of the examples there and maybe change some of the wordings to make it better to read.

  app: {
    on: jest.fn(),
    whenReady: jest.fn(() => Promise.resolve()),
  },
  BrowserWindow: jest.fn().mockImplementation(() => ({
    // partial mocks.
  }))
}));

const { BrowserWindow } = require('electron');

> const exported = require('main.cjs');

Reviewer 2: ```suggestion
const exported = require('./main.cjs');
```

// jest === import.meta.jest => true
```

Additionally, since ESM evaluates static `import` statements before looking at the code, hoisting on `jest.mock` calls that happens in CJS modules won't work in ESM. To `mock` modules in ESM, you need to use `require` or dynamic `import()` after `jest.mock` calls to load the mocked modules, same applies to modules which have to load the mocked modules.

> Please also note that we currently don't support `jest.mock` in a clean way in ESM, but that is something we intend to add proper support for in the future. Follow [this issue](https://github.com/facebook/jest/issues/10025) for updates. `jest.unstable_mockModule` is needed to mock ESM for now. Its usage is essentially the same as `jest.mock`. See the example below:

Reviewer: If I get it right, `jest.unstable_mockModule` if the effort to support mocking in ESM:

```suggestion
ESM module mocking is supported through `jest.unstable_mockModule`. As the name suggest works are in progress, please follow [this issue](https://github.com/facebook/jest/issues/10025) for updates.

The usage of `jest.unstable_mockModule` is essentially the same as `jest.mock` with two differences: the factory function is required and it can be sync or async:
```

```js
import {jest} from '@jest/globals';

jest.unstable_mockModule('node:child_process', () => ({
  execSync: jest.fn(),
  // etc.
}));

const {execSync} = await import('node:child_process');

// etc.
```

For mocking CJS modules, you should continue to use `jest.mock`. See the example below:

```js title="main.cjs"
const {BrowserWindow, app} = require('electron');

// etc.

module.exports = {example};
```

> <!-- eslint-disable no-redeclare -->

Reviewer: Just add an `override` here: https://github.com/facebook/jest/blob/2fa8763db2e661b520ebebeef45a22897d6da673/.eslintrc.cjs#L183-L189

Author: Okay

// jest === import.meta.jest => true
```

> Additionally, since ESM evaluates static `import` statements before looking at the code, hoisting on `jest.mock` calls that happens in CJS modules won't work in ESM. To `mock` modules in ESM, you need to use `require` or dynamic `import()` after `jest.mock` calls to load the mocked modules, same applies to modules which have to load the mocked modules.

Reviewer: Last idea from my side:

```suggestion

## Module mocking in ESM

Since ESM evaluates static `import` statements before looking at the code, hoisting on `jest.mock` calls that happens in CJS modules won't work in ESM. To `mock` modules in ESM, you need to use `require` or dynamic `import()` after `jest.mock` calls to load the mocked modules, same applies to modules which have to load the mocked modules.
```

Author: How about we have this title added on v27+ only? We can only `mock` CJS modules before v27 even with ESM support enabled right? Plus I've noticed that the `ECMAScript Modules` page doesn't show up on v25's sidebar. Is that list a generated one or we have to update the `sidebars.json` manually?

Reviewer: Right, it makes sense only with v27+. I will take a look at the sidebars later. Can it be there was no ESM support in v25? No sure.

Reviewer: Looked around quickly. Seems like first steps with ESM support were released in v25. So you are right here too, that link is missing in the sidebar and must be added. Thanks!

// jest === import.meta.jest => true
```

## Module mocking in ESM

> Since ESM evaluates static `import` statements before looking at the code, hoisting on `jest.mock` calls that happens in CJS modules won't work in ESM. To `mock` modules in ESM, you need to use `require` or dynamic `import()` after `jest.mock` calls to load the mocked modules, same applies to modules which have to load the mocked modules.

Reviewer 2: ```suggestion
Since ESM evaluates static `import` statements before looking at the code, the hoisting of `jest.mock` calls that happens in CJS won't work for ESM. To mock modules in ESM, you need to use `require` or dynamic `import()` after `jest.mock` calls to load the mocked modules - the same applies to modules which load the mocked modules.
```

> ESM module mocking is supported through `jest.unstable_mockModule`. As the name suggest works are in progress, please follow [this issue](https://github.com/facebook/jest/issues/10025) for updates.

Reviewer 2: ```suggestion
ESM mocking is supported through `jest.unstable_mockModule`. As the name suggests, this API is still work in progress, please follow [this issue](https://github.com/facebook/jest/issues/10025) for updates.
```
